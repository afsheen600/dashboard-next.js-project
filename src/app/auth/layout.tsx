// app/(auth)/layout.tsx
"use client";
import { useEffect, useState } from "react";
import { useRouter } from "next/navigation";
import { auth } from "@/app/services/firebaseConfig";
import { onAuthStateChanged } from "firebase/auth";
import { SidebarProvider } from "@/context/SidebarContext";
// import Navbar from "@/components/Navbar";
import Navbar from "../components/Navbar";
import Sidebar from "../components/Sidebar";

export default function AuthLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  const router = useRouter();
  const [authenticated, setAuthenticated] = useState(false);

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (user) => {
      if (!user) {
        router.push("/login");
      } else {
        setAuthenticated(true);
      }
    });
    return () => unsubscribe();
  }, [router]);

  if (!authenticated) {
    return <div>Loading...</div>; // Or a proper loading spinner
  }

  return (
    <SidebarProvider>
      <div style={{ display: "flex" }}>
        <Navbar />
        <Sidebar />
        <main style={{ flexGrow: 1, padding: "24px", marginTop: "64px" }}>
          {children}
        </main>
      </div>
    </SidebarProvider>
  );
}
