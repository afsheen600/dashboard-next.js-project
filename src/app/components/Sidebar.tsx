"use client";

import Link from "next/link";
import {
  Drawer,
  List,
  ListItem,
  ListItemButton,
  ListItemIcon,
  ListItemText,
  Toolbar,
  Divider,
} from "@mui/material";
import DashboardIcon from "@mui/icons-material/Dashboard";
import InfoIcon from "@mui/icons-material/Info";
import AnalyticsIcon from "@mui/icons-material/Analytics";
import ShoppingCartIcon from "@mui/icons-material/ShoppingCart";
import LogoutIcon from "@mui/icons-material/Logout";
import { useSidebar } from "../../context/SidebarContext";
import { auth } from "@/app/services/firebaseConfig";
import { signOut } from "firebase/auth";
import { useRouter } from "next/navigation";

interface MenuItem {
  text: string;
  path: string;
  icon: React.ReactNode;
}

const Sidebar: React.FC = () => {
  const {
    isOpen,
    toggleSidebar,
  }: {
    isOpen: boolean;
    toggleSidebar: () => void;
  } = useSidebar();

  const router = useRouter();

  const handleLogout = async () => {
    try {
      await signOut(auth);
      router.push("/login");
    } catch (error) {
      console.error("Error signing out: ", error);
    }
  };

  const menuItems: MenuItem[] = [
    { text: "Dashboard", path: "/", icon: <DashboardIcon /> },
    {
      text: "Analytics",
      path: "/pages/dashboard/analytics",
      icon: <AnalyticsIcon />,
    },
    {
      text: "ProductsTable",
      path: "/pages/dashboard/products",
      icon: <ShoppingCartIcon />,
    },
    { text: "Settings", path: "/pages/dashboard/settings", icon: <InfoIcon /> },
  ];

  return (
    <Drawer
      variant="temporary"
      open={isOpen}
      onClose={toggleSidebar}
      ModalProps={{
        keepMounted: true,
      }}
      sx={{
        "& .MuiDrawer-paper": {
          boxSizing: "border-box",
          width: 240,
          display: "flex",
          flexDirection: "column",
          justifyContent: "space-between",
        },
      }}
    >
      <div>
        <Toolbar />
        <List>
          {menuItems.map((item) => (
            <ListItem key={item.text} disablePadding>
              <Link
                href={item.path}
                style={{
                  textDecoration: "none",
                  color: "inherit",
                  width: "100%",
                }}
              >
                <ListItemButton onClick={toggleSidebar}>
                  <ListItemIcon>{item.icon}</ListItemIcon>
                  <ListItemText primary={item.text} />
                </ListItemButton>
              </Link>
            </ListItem>
          ))}
        </List>
      </div>
      <div>
        <Divider />
        <List>
          <ListItem disablePadding>
            <ListItemButton onClick={handleLogout}>
              <ListItemIcon>
                <LogoutIcon />
              </ListItemIcon>
              <ListItemText primary="Logout" />
            </ListItemButton>
          </ListItem>
        </List>
      </div>
    </Drawer>
  );
};

export default Sidebar;
