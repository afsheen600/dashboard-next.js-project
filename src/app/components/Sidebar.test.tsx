import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import Sidebar from "./Sidebar";
import { signOut } from "firebase/auth";

// Mock useSidebar context
const toggleSidebarMock = jest.fn();
jest.mock("../../context/SidebarContext", () => ({
  useSidebar: () => ({
    isOpen: true,
    toggleSidebar: toggleSidebarMock,
  }),
}));

// Mock next/navigation
const pushMock = jest.fn();
jest.mock("next/navigation", () => ({
  useRouter: () => ({
    push: pushMock,
  }),
  usePathname: () => "/",
}));

// Mock firebase/auth
jest.mock("firebase/auth", () => {
  const signOutMock = jest.fn(() => Promise.resolve());
  return {
    getAuth: jest.fn(() => ({
      currentUser: { uid: "test-user" },
      signOut: signOutMock,
    })),
    signOut: signOutMock,
  };
});

describe("Sidebar", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders all menu items", () => {
    render(<Sidebar />);
    expect(screen.getByText(/dashboard/i)).toBeInTheDocument();
    expect(screen.getByText(/analytics/i)).toBeInTheDocument();
    expect(screen.getByText(/products/i)).toBeInTheDocument();
    expect(screen.getByText(/settings/i)).toBeInTheDocument();
  });

  it("renders the logout button", () => {
    render(<Sidebar />);
    expect(screen.getByText(/logout/i)).toBeInTheDocument();
  });

  it("calls signOut and redirects on logout click", async () => {
    render(<Sidebar />);
    fireEvent.click(screen.getByText(/logout/i));
    expect(signOut).toHaveBeenCalled();
    expect(pushMock).toHaveBeenCalledWith("/login");
    expect(pushMock).toHaveBeenCalledWith("/login");
  });
});

it("calls toggleSidebar when menu item is clicked", () => {
  render(<Sidebar />);
  fireEvent.click(screen.getByText(/dashboard/i));
  expect(toggleSidebarMock).toHaveBeenCalled();
});

it("matches snapshot", () => {
  const { asFragment } = render(<Sidebar />);
  expect(asFragment()).toMatchSnapshot();
});
