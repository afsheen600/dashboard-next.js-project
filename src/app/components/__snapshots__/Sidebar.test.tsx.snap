// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Sidebar matches snapshot 1`] = `<DocumentFragment />`;

exports[`matches snapshot 1`] = `<DocumentFragment />`;
